nautobot/utilities/filters.py

```diff
@@ -415,6 +416,14 @@ def filter(self, qs, value):
         return qs.distinct()
 
 
+class NaturalKeyMultipleChoiceFilter(django_filters.ModelMultipleChoiceFilter):
+    """
+    Filter that supports filtering on the `pk` or the `slug` of a foreign-key related field.
```

Comment:
Does this pattern work to implement filters that use nested lookups as well, such as we see in `circuits`?

```python
    site_id = django_filters.ModelMultipleChoiceFilter(
        field_name="circuits__terminations__site",
        queryset=Site.objects.all(),
        label="Site",
    )
    site = django_filters.ModelMultipleChoiceFilter(
        field_name="circuits__terminations__site__slug",
        queryset=Site.objects.all(),
        to_field_name="slug",
        label="Site (slug)",
    )
```

Reply:
It does! I tested it on CircuitFilterSet:

```py
    site = SlugOrPKMultipleChoiceFilter(
        field_name="terminations__site",
        queryset=Site.objects.all(),
        label="Site (Slug or ID)",
    )
```

<details>
  <summary>All circuits</summary>

```
GET /api/circuits/circuits/
HTTP 200 OK
API-Version: 1.2
Allow: GET, POST, PUT, PATCH, DELETE, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept
```
```json
{
    "count": 2,
    "next": null,
    "previous": null,
    "results": [
        {
            "id": "2c2d4a1c-24ef-4483-8f70-dd907a342130",
            "url": "http://localhost:8080/api/circuits/circuits/2c2d4a1c-24ef-4483-8f70-dd907a342130/",
            "cid": "CCSD123456",
            "provider": {
                "id": "7f725747-b88d-40d9-ba90-b35cd365aa4d",
                "url": "http://localhost:8080/api/circuits/providers/7f725747-b88d-40d9-ba90-b35cd365aa4d/",
                "name": "TMO",
                "slug": "tmo",
                "display": "TMO"
            },
            "type": {
                "id": "9b7902e6-aa2e-4e3c-a2c8-6092e2f4f9f5",
                "url": "http://localhost:8080/api/circuits/circuit-types/9b7902e6-aa2e-4e3c-a2c8-6092e2f4f9f5/",
                "name": "Ethernet",
                "slug": "ethernet",
                "display": "Ethernet"
            },
            "status": {
                "value": "active",
                "label": "Active"
            },
            "tenant": null,
            "install_date": "2022-06-07",
            "commit_rate": 2048,
            "description": "",
            "termination_a": {
                "id": "acead8ad-1e40-43f7-9085-37030bad848e",
                "url": "http://localhost:8080/api/circuits/circuit-terminations/acead8ad-1e40-43f7-9085-37030bad848e/",
                "site": {
                    "id": "20c70383-83d2-4230-a300-960a0eb2d06c",
                    "url": "http://localhost:8080/api/dcim/sites/20c70383-83d2-4230-a300-960a0eb2d06c/",
                    "name": "ALG01",
                    "slug": "alg01",
                    "display": "ALG01"
                },
                "provider_network": null,
                "connected_endpoint": null,
                "port_speed": null,
                "upstream_speed": null,
                "xconnect_id": "",
                "display": "Termination A: ALG01"
            },
            "termination_z": {
                "id": "741e73ef-9d10-4c98-9e3d-40e2e528793b",
                "url": "http://localhost:8080/api/circuits/circuit-terminations/741e73ef-9d10-4c98-9e3d-40e2e528793b/",
                "site": {
                    "id": "f4c20fd4-cb0f-40ad-83a3-e4c788bcd9be",
                    "url": "http://localhost:8080/api/dcim/sites/f4c20fd4-cb0f-40ad-83a3-e4c788bcd9be/",
                    "name": "DEL01",
                    "slug": "del01",
                    "display": "DEL01"
                },
                "provider_network": null,
                "connected_endpoint": null,
                "port_speed": null,
                "upstream_speed": null,
                "xconnect_id": "",
                "display": "Termination Z: DEL01"
            },
            "comments": "",
            "tags": [],
            "custom_fields": {},
            "created": "2022-06-07",
            "last_updated": "2022-06-07T14:20:06.550225Z",
            "display": "CCSD123456"
        },
        {
            "id": "7fdd04a8-d7d3-4045-9bf0-14910f9136fc",
            "url": "http://localhost:8080/api/circuits/circuits/7fdd04a8-d7d3-4045-9bf0-14910f9136fc/",
            "cid": "CCSD789456123",
            "provider": {
                "id": "7f725747-b88d-40d9-ba90-b35cd365aa4d",
                "url": "http://localhost:8080/api/circuits/providers/7f725747-b88d-40d9-ba90-b35cd365aa4d/",
                "name": "TMO",
                "slug": "tmo",
                "display": "TMO"
            },
            "type": {
                "id": "9b7902e6-aa2e-4e3c-a2c8-6092e2f4f9f5",
                "url": "http://localhost:8080/api/circuits/circuit-types/9b7902e6-aa2e-4e3c-a2c8-6092e2f4f9f5/",
                "name": "Ethernet",
                "slug": "ethernet",
                "display": "Ethernet"
            },
            "status": {
                "value": "planned",
                "label": "Planned"
            },
            "tenant": null,
            "install_date": null,
            "commit_rate": null,
            "description": "",
            "termination_a": {
                "id": "a4a6dbf1-02cb-4a53-b950-5af5ad229a5a",
                "url": "http://localhost:8080/api/circuits/circuit-terminations/a4a6dbf1-02cb-4a53-b950-5af5ad229a5a/",
                "site": {
                    "id": "38ee7a4c-32a7-4127-aca5-2362b5679054",
                    "url": "http://localhost:8080/api/dcim/sites/38ee7a4c-32a7-4127-aca5-2362b5679054/",
                    "name": "ATL01",
                    "slug": "atl01",
                    "display": "ATL01"
                },
                "provider_network": null,
                "connected_endpoint": null,
                "port_speed": null,
                "upstream_speed": null,
                "xconnect_id": "",
                "display": "Termination A: ATL01"
            },
            "termination_z": {
                "id": "18a85838-fd1d-4bf0-80c3-6cfbc77c64e3",
                "url": "http://localhost:8080/api/circuits/circuit-terminations/18a85838-fd1d-4bf0-80c3-6cfbc77c64e3/",
                "site": {
                    "id": "de71e1db-d0fa-49b8-a8e4-f9e3839c0bc8",
                    "url": "http://localhost:8080/api/dcim/sites/de71e1db-d0fa-49b8-a8e4-f9e3839c0bc8/",
                    "name": "ATL02",
                    "slug": "atl02",
                    "display": "ATL02"
                },
                "provider_network": null,
                "connected_endpoint": null,
                "port_speed": null,
                "upstream_speed": null,
                "xconnect_id": "",
                "display": "Termination Z: ATL02"
            },
            "comments": "",
            "tags": [],
            "custom_fields": {},
            "created": "2022-06-14",
            "last_updated": "2022-06-14T14:12:13.419635Z",
            "display": "CCSD789456123"
        }
    ]
}
```

</details>

<details>
  <summary>Filter by slug</summary>

```
GET /api/circuits/circuits/?site=atl01
HTTP 200 OK
API-Version: 1.2
Allow: GET, POST, PUT, PATCH, DELETE, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept
```
```json
{
    "count": 1,
    "next": null,
    "previous": null,
    "results": [
        {
            "id": "7fdd04a8-d7d3-4045-9bf0-14910f9136fc",
            "url": "http://localhost:8080/api/circuits/circuits/7fdd04a8-d7d3-4045-9bf0-14910f9136fc/",
            "cid": "CCSD789456123",
            "provider": {
                "id": "7f725747-b88d-40d9-ba90-b35cd365aa4d",
                "url": "http://localhost:8080/api/circuits/providers/7f725747-b88d-40d9-ba90-b35cd365aa4d/",
                "name": "TMO",
                "slug": "tmo",
                "display": "TMO"
            },
            "type": {
                "id": "9b7902e6-aa2e-4e3c-a2c8-6092e2f4f9f5",
                "url": "http://localhost:8080/api/circuits/circuit-types/9b7902e6-aa2e-4e3c-a2c8-6092e2f4f9f5/",
                "name": "Ethernet",
                "slug": "ethernet",
                "display": "Ethernet"
            },
            "status": {
                "value": "planned",
                "label": "Planned"
            },
            "tenant": null,
            "install_date": null,
            "commit_rate": null,
            "description": "",
            "termination_a": {
                "id": "a4a6dbf1-02cb-4a53-b950-5af5ad229a5a",
                "url": "http://localhost:8080/api/circuits/circuit-terminations/a4a6dbf1-02cb-4a53-b950-5af5ad229a5a/",
                "site": {
                    "id": "38ee7a4c-32a7-4127-aca5-2362b5679054",
                    "url": "http://localhost:8080/api/dcim/sites/38ee7a4c-32a7-4127-aca5-2362b5679054/",
                    "name": "ATL01",
                    "slug": "atl01",
                    "display": "ATL01"
                },
                "provider_network": null,
                "connected_endpoint": null,
                "port_speed": null,
                "upstream_speed": null,
                "xconnect_id": "",
                "display": "Termination A: ATL01"
            },
            "termination_z": {
                "id": "18a85838-fd1d-4bf0-80c3-6cfbc77c64e3",
                "url": "http://localhost:8080/api/circuits/circuit-terminations/18a85838-fd1d-4bf0-80c3-6cfbc77c64e3/",
                "site": {
                    "id": "de71e1db-d0fa-49b8-a8e4-f9e3839c0bc8",
                    "url": "http://localhost:8080/api/dcim/sites/de71e1db-d0fa-49b8-a8e4-f9e3839c0bc8/",
                    "name": "ATL02",
                    "slug": "atl02",
                    "display": "ATL02"
                },
                "provider_network": null,
                "connected_endpoint": null,
                "port_speed": null,
                "upstream_speed": null,
                "xconnect_id": "",
                "display": "Termination Z: ATL02"
            },
            "comments": "",
            "tags": [],
            "custom_fields": {},
            "created": "2022-06-14",
            "last_updated": "2022-06-14T14:12:13.419635Z",
            "display": "CCSD789456123"
        }
    ]
}
```

</details>

<details>
  <summary>Filter by pk</summary>
  
```
GET /api/circuits/circuits/?site=f4c20fd4-cb0f-40ad-83a3-e4c788bcd9be
HTTP 200 OK
API-Version: 1.2
Allow: GET, POST, PUT, PATCH, DELETE, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept
```
```json
{
    "count": 1,
    "next": null,
    "previous": null,
    "results": [
        {
            "id": "2c2d4a1c-24ef-4483-8f70-dd907a342130",
            "url": "http://localhost:8080/api/circuits/circuits/2c2d4a1c-24ef-4483-8f70-dd907a342130/",
            "cid": "CCSD123456",
            "provider": {
                "id": "7f725747-b88d-40d9-ba90-b35cd365aa4d",
                "url": "http://localhost:8080/api/circuits/providers/7f725747-b88d-40d9-ba90-b35cd365aa4d/",
                "name": "TMO",
                "slug": "tmo",
                "display": "TMO"
            },
            "type": {
                "id": "9b7902e6-aa2e-4e3c-a2c8-6092e2f4f9f5",
                "url": "http://localhost:8080/api/circuits/circuit-types/9b7902e6-aa2e-4e3c-a2c8-6092e2f4f9f5/",
                "name": "Ethernet",
                "slug": "ethernet",
                "display": "Ethernet"
            },
            "status": {
                "value": "active",
                "label": "Active"
            },
            "tenant": null,
            "install_date": "2022-06-07",
            "commit_rate": 2048,
            "description": "",
            "termination_a": {
                "id": "acead8ad-1e40-43f7-9085-37030bad848e",
                "url": "http://localhost:8080/api/circuits/circuit-terminations/acead8ad-1e40-43f7-9085-37030bad848e/",
                "site": {
                    "id": "20c70383-83d2-4230-a300-960a0eb2d06c",
                    "url": "http://localhost:8080/api/dcim/sites/20c70383-83d2-4230-a300-960a0eb2d06c/",
                    "name": "ALG01",
                    "slug": "alg01",
                    "display": "ALG01"
                },
                "provider_network": null,
                "connected_endpoint": null,
                "port_speed": null,
                "upstream_speed": null,
                "xconnect_id": "",
                "display": "Termination A: ALG01"
            },
            "termination_z": {
                "id": "741e73ef-9d10-4c98-9e3d-40e2e528793b",
                "url": "http://localhost:8080/api/circuits/circuit-terminations/741e73ef-9d10-4c98-9e3d-40e2e528793b/",
                "site": {
                    "id": "f4c20fd4-cb0f-40ad-83a3-e4c788bcd9be",
                    "url": "http://localhost:8080/api/dcim/sites/f4c20fd4-cb0f-40ad-83a3-e4c788bcd9be/",
                    "name": "DEL01",
                    "slug": "del01",
                    "display": "DEL01"
                },
                "provider_network": null,
                "connected_endpoint": null,
                "port_speed": null,
                "upstream_speed": null,
                "xconnect_id": "",
                "display": "Termination Z: DEL01"
            },
            "comments": "",
            "tags": [],
            "custom_fields": {},
            "created": "2022-06-07",
            "last_updated": "2022-06-07T14:20:06.550225Z",
            "display": "CCSD123456"
        }
    ]
}
```

</details>